incubator/hnc/pkg/kubectl/set.go

```diff
@@ -16,100 +16,148 @@ limitations under the License.
 package kubectl
 
 import (
+	"errors"
 	"fmt"
 	"os"
 
 	"github.com/spf13/cobra"
+
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
 )
 
+//flagValues struct stores name of namespaces against type of flag passed
+type flagValues struct {
+	root                bool
+	parentString        string
+	requiredChildArray  []string
+	optionalChildString string
+}
+
 var setCmd = &cobra.Command{
 	Use:     "set",
 	Short:   "Allows setting namespace hierarchy",
 	Aliases: []string{"set"},
 	Args:    cobra.MinimumNArgs(1),
 	Run: func(cmd *cobra.Command, args []string) {
 		nnm := args[0]
-		hc := client.getHierarchy(nnm)
-		oldpnm := hc.Spec.Parent
 		flags := cmd.Flags()
-		numChanges := 0
+		parentString, _ := flags.GetString("parent")
+		requiredChildArray, _ := flags.GetStringArray("requiredChild")
+		optionalChildString, _ := flags.GetString("optionalChild")
+
+		flagValues := flagValues{
+			root:                flags.Changed("root"),
+			parentString:        parentString,
+			requiredChildArray:  requiredChildArray,
+			optionalChildString: optionalChildString,
+		}
 
-		if flags.Changed("root") && flags.Changed("parent") {
-			fmt.Println("Cannot set both --root and --parent at the same time")
+		msgs, err := setCmdFunc(client, flagValues, nnm)
+		if err != nil {
+			fmt.Println(err.Error())
 			os.Exit(1)
 		}
 
-		if flags.Changed("root") {
-			if oldpnm == "" {
-				fmt.Printf("%s is already a root namespace; unchanged", nnm)
-			} else {
-				hc.Spec.Parent = ""
-				fmt.Printf("Unsetting the parent of %s (was previously %s)\n", nnm, oldpnm)
-				numChanges++
-			}
+		for _, each := range msgs {
+			fmt.Print(each)
 		}
 
-		if flags.Changed("parent") {
-			pnm, _ := flags.GetString("parent")
-			if oldpnm == pnm {
-				fmt.Printf("Parent of %s is already %s; unchanged\n", nnm, pnm)
-			} else {
-				hc.Spec.Parent = pnm
-				if oldpnm == "" {
-					fmt.Printf("Setting the parent of %s to %s\n", nnm, pnm)
-				} else {
-					fmt.Printf("Changing the parent of %s from %s to %s\n", nnm, oldpnm, pnm)
-				}
-				numChanges++
-			}
+	},
+}
+
+func setCmdFunc(cl Client, flagValues flagValues, nnm string) ([]string, error) {
+	msg := make([]string, 0)
+	hc := cl.getHierarchy(nnm)
+	oldpnm := hc.Spec.Parent
+	numChanges := 0
+
+	if flagValues.root && flagValues.parentString != "" {
+		return msg, errors.New("Cannot set both --root and --parent at the same time \n")
+	}
+
+	if flagValues.root {
+		setRoot(hc, oldpnm, nnm, &msg, &numChanges)
+	}
+
+	if flagValues.parentString != "" {
+		setParent(hc, oldpnm, flagValues.parentString, nnm, &msg, &numChanges)
+	}
+
+	if flagValues.requiredChildArray != nil {
```

Comment:
It's preferred to say `if len(foo) != 0`, although in this particular case, `setRequiredChildren` should probably work even if you pass in an empty array.